src/mpi/datatype/typeutil.c

```diff
@@ -26,90 +28,97 @@ static int datatype_attr_finalize_cb(void *dummy);
 
 typedef struct mpi_names_t {
     MPI_Datatype dtype;
+    int category;               /* MPIR_TYPE_{FIXED,SIGNED,UNSIGNED,FLOAT,COMPLEX} */
     const char *name;
 } mpi_names_t;
-#define type_name_entry(x_) { x_, #x_ }
+#define type_name_entry(x_, category) { x_, category, #x_ }
+#define type_name_entry_fixed(x_) { x_, MPIR_TYPE_FIXED, #x_ }
+#define type_name_entry_signed(x_) { x_, MPIR_TYPE_SIGNED, #x_ }
+#define type_name_entry_unsigned(x_) { x_, MPIR_TYPE_UNSIGNED, #x_ }
+#define type_name_entry_float(x_) { x_, MPIR_TYPE_FLOAT, #x_ }
+#define type_name_entry_complex(x_) { x_, MPIR_TYPE_COMPLEX, #x_ }
 
 static mpi_names_t mpi_dtypes[] = {
-    type_name_entry(MPI_CHAR),
-    type_name_entry(MPI_UNSIGNED_CHAR),
-    type_name_entry(MPI_SIGNED_CHAR),
-    type_name_entry(MPI_BYTE),
-    type_name_entry(MPI_WCHAR),
-    type_name_entry(MPI_SHORT),
-    type_name_entry(MPI_UNSIGNED_SHORT),
-    type_name_entry(MPI_INT),
-    type_name_entry(MPI_UNSIGNED),
-    type_name_entry(MPI_LONG),
-    type_name_entry(MPI_UNSIGNED_LONG),
-    type_name_entry(MPI_FLOAT),
-    type_name_entry(MPI_DOUBLE),
-    type_name_entry(MPI_LONG_DOUBLE),
-    type_name_entry(MPI_LONG_LONG_INT),
-    type_name_entry(MPI_UNSIGNED_LONG_LONG),
-    type_name_entry(MPI_PACKED),
-    type_name_entry(MPI_LB),
-    type_name_entry(MPI_UB),
-    type_name_entry(MPI_2INT),
+    type_name_entry_signed(MPI_CHAR),
```

Comment:
IIRC, The signedness of `char` depends on the platform. Not sure if that observation is relevant here, but you may want to treat `char` as "fixed" instead?.

Reply:
Or maybe just treat `char` as opaque bytes? IIRC, `MPI_CHAR` is not intended to work with reductions, therefore what's the point of tracking or dealing with `char` signedness within MPI internals?

Reply:
All these `type_name_entry` probably will be removed and we'll directly use configure to map the internal entries.

Reply:
BTW, I think people use `char` for arithmetic all the time and thus we need support `MPI_CHAR` for reduction.

Reply:
Yes, people use `char` for arith, most likely for legacy reasons and unaware of the platform dependence of its signedness. People should really be using `signed/unsigned` `char` explicitly for arithmetics.

As far as I can tell, the MPI standard does not allow reductions with `MPI_CHAR`:
- MPI 4.1, page 227: `MPI_CHAR` is not in the list of datatypes intended for reductions, but the signed/unsigned char variants are.
- MPI 4.1, page 1067: see the comment *MPI_CHAR is not allowed for predefined reduction operations.*

Reply:
Okay, I'll take your advice and map `MPI_CHAR` to `MPIR_FIXED1`, which raises error on reduction.